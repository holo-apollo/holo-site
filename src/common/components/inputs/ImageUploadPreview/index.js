// @flow
import * as React from 'react';

import Attach from 'common/components/icons/Attach';
import Close from 'common/components/icons/Close';
import FieldWithError from '../FieldWithError';
import {
  LabelTextCont,
  StyledLabel,
  StyledInput,
  ButtonTextCont,
  HelpTextCont,
  ImagesCont,
  ImageCont,
  StyledImg,
  CloseCont,
} from './styled';

type FileChoiceEvent = SyntheticInputEvent<HTMLInputElement>;

type Props = {
  name: string,
  label: string,
  buttonText: string,
  helperText: string | React.Node,
  errorText?: string,
  onChange?: FileChoiceEvent => void,
  onRemove?: File => void,
};

type Image = {
  file: File,
  previewUrl: string | ArrayBuffer,
};

const ImageUploadPreview = (props: Props) => {
  const [images, setImages] = React.useState([]);

  function handleChange(event: FileChoiceEvent) {
    event.preventDefault();

    const files = event.target.files;
    for (let i = 0; i < files.length; i++) {
      const file = files[i];
      if (file.type.startsWith('image/')) {
        const reader = new FileReader();
        reader.onloadend = () => {
          const image = {
            file,
            previewUrl: reader.result,
          };
          setImages([...images, image]);
        };
        reader.readAsDataURL(file);
      }
    }

    props.onChange && props.onChange(event);
  }

  function handleImageRemove(image: Image) {
    setImages(images.filter(item => item.file !== image.file));
    props.onRemove && props.onRemove(image.file);
  }

  function renderImagePreview(image: Image, index: number) {
    return (
      <ImageCont key={index}>
        <StyledImg src={image.previewUrl} />
        <CloseCont onClick={() => handleImageRemove(image)}>
          <Close height={13} />
        </CloseCont>
      </ImageCont>
    );
  }

  return (
    <FieldWithError errorText={props.errorText}>
      <LabelTextCont error={Boolean(props.errorText)}>
        {props.label}
      </LabelTextCont>
      <StyledLabel>
        <StyledInput
          type="file"
          multiple={true}
          name={props.name}
          onChange={handleChange}
        />
        <ButtonTextCont>
          <Attach height={18} />
          {props.buttonText}
        </ButtonTextCont>
      </StyledLabel>
      <HelpTextCont>{props.helperText}</HelpTextCont>
      <ImagesCont>{images.map(renderImagePreview)}</ImagesCont>
    </FieldWithError>
  );
};

export default ImageUploadPreview;
