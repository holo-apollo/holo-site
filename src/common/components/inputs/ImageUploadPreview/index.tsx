import React, { useState, ChangeEvent, ReactNode } from 'react';

import Attach from 'common/components/icons/Attach';
import Close from 'common/components/icons/Close';
import FieldWithError from '../FieldWithError';
import {
  LabelTextCont,
  StyledLabel,
  StyledInput,
  ButtonTextCont,
  HelpTextCont,
  ImagesCont,
  ImageCont,
  StyledImg,
  CloseCont,
} from './styled';

type FileChoiceEvent = ChangeEvent<HTMLInputElement>;

type Props = {
  name: string;
  label: string;
  buttonText: string;
  helperText: ReactNode;
  errorText?: string;
  onChange?: (event: FileChoiceEvent) => void;
  onRemove?: (file: File) => void;
};

type Image = {
  file: File;
  previewUrl: string;
};

const ImageUploadPreview = (props: Props) => {
  const [images, setImages] = useState<Image[]>([]);

  function handleChange(event: FileChoiceEvent) {
    event.preventDefault();

    const files = event.target.files || [];
    for (let i = 0; i < files.length; i++) {
      const file = files[i];
      if (file.type.startsWith('image/')) {
        const reader = new FileReader();
        reader.onloadend = () => {
          const image: Image = {
            file,
            previewUrl: reader.result === null ? '' : reader.result.toString(),
          };
          setImages([...images, image]);
        };
        reader.readAsDataURL(file);
      }
    }

    props.onChange && props.onChange(event);
  }

  function handleImageRemove(image: Image) {
    setImages(images.filter(item => item.file !== image.file));
    props.onRemove && props.onRemove(image.file);
  }

  function renderImagePreview(image: Image, index: number) {
    return (
      <ImageCont key={index}>
        <StyledImg src={image.previewUrl} />
        <CloseCont onClick={() => handleImageRemove(image)}>
          <Close height={13} />
        </CloseCont>
      </ImageCont>
    );
  }

  return (
    <FieldWithError errorText={props.errorText}>
      <LabelTextCont error={Boolean(props.errorText)}>
        {props.label}
      </LabelTextCont>
      <StyledLabel>
        <StyledInput
          type="file"
          multiple={true}
          name={props.name}
          onChange={handleChange}
        />
        <ButtonTextCont>
          <Attach height={18} />
          {props.buttonText}
        </ButtonTextCont>
      </StyledLabel>
      <HelpTextCont>{props.helperText}</HelpTextCont>
      <ImagesCont>{images.map(renderImagePreview)}</ImagesCont>
    </FieldWithError>
  );
};

export default ImageUploadPreview;
