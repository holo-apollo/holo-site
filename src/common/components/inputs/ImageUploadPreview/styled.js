import styled, { css } from 'styled-components';

import palette from 'common/palette';

const getLabelStyles = ({ error }) => css`
  color: ${error ? palette.red : palette.grey};
`;

export const LabelTextCont = styled.div`
  font-size: 12px;
  margin-bottom: 10px;
  text-align: left;

  ${getLabelStyles};
`;

export const StyledLabel = styled.label`
  color: ${palette.blue};
  cursor: pointer;

  &:hover {
    color: ${palette.deepBlue};

    svg {
      fill: ${palette.deepBlue};

      g {
        fill: ${palette.deepBlue};
      }
    }
  }
`;

export const StyledInput = styled.input`
  display: none;
`;

export const ButtonTextCont = styled.div`
  display: flex;
  font-size: 12px;

  svg {
    margin-right: 5px;
  }
`;

export const HelpTextCont = styled.div`
  font-size: 10px;
  color: ${palette.grey};
  text-align: left;
`;

export const ImagesCont = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin-top: 25px;
`;

export const ImageCont = styled.div`
  margin-right: 20px;
  margin-bottom: 20px;
  position: relative;

  @media (min-width: 460px) {
    &:nth-child(4n) {
      margin-right: 0;
    }
  }

  @media (min-width: 350px) and (max-width: 459px) {
    &:nth-child(3n) {
      margin-right: 0;
    }
  }

  @media (max-width: 349px) {
    &:nth-child(2n) {
      margin-right: 0;
    }
  }
`;

export const StyledImg = styled.img`
  width: 88px;
  height: 88px;
  object-fit: cover;
  border: 1px solid ${palette.darkGrey};
`;

export const CloseCont = styled.span`
  position: absolute;
  top: 5px;
  right: 5px;
  cursor: pointer;
`;
