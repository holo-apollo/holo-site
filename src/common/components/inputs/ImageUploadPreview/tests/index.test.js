import React from 'react';
import renderer from 'react-test-renderer';

import ImageUploadPreview from '../index';

describe('ImageUploadPreview', function() {
  it('should match snapshot', function() {
    const tree = renderer
      .create(
        <ImageUploadPreview
          name="name"
          label="Label"
          buttonText="Button text"
          helperText="Helper text"
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
