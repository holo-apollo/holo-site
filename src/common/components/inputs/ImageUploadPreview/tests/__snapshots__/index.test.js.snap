// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageUploadPreview should match snapshot 1`] = `
Array [
  <div
    className="styled__LabelTextCont-sc-3qf63z-0 flMOyl"
  >
    Label
  </div>,
  <label
    className="styled__StyledLabel-sc-3qf63z-1 bhlSvV"
  >
    <input
      className="styled__StyledInput-sc-3qf63z-2 bBdwie"
      multiple={true}
      name="name"
      onChange={[Function]}
      type="file"
    />
    <div
      className="styled__ButtonTextCont-sc-3qf63z-3 bxNgpQ"
    >
      <svg
        className="SVGIcon__StyledSVG-sc-15q48pi-0 iedNHS"
        fill="#4caad0"
        height={18}
        viewBox="0 0 18 18"
        width={18}
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="#4caad0"
          transform="translate(-750.000000, -918.000000)"
        >
          <path
            d="M756.086144,936 C754.458092,936 752.931635,935.37507 751.781078,934.238148 C749.406307,931.891523 749.406307,928.074177 751.781078,925.730061 L758.313602,919.274958 C759.146677,918.451757 760.254057,918 761.432553,918 C762.611048,918 763.718428,918.454267 764.551503,919.274958 C766.270989,920.974066 766.270989,923.737312 764.551503,925.436419 L758.613306,931.304239 C758.097714,931.81372 757.411951,932.094813 756.683011,932.094813 C755.954071,932.094813 755.268308,931.81372 754.752716,931.304239 C753.688514,930.252649 753.688514,928.540993 754.752716,927.489403 L759.801961,922.50251 C760.048328,922.259063 760.447086,922.259063 760.693453,922.50251 C760.939819,922.745956 760.939819,923.139989 760.693453,923.383436 L755.644208,928.370329 C755.070199,928.937535 755.070199,929.858617 755.644208,930.423313 C756.197898,930.97295 757.168125,930.97295 757.721815,930.423313 L763.660011,924.555494 C764.886764,923.34077 764.886764,921.368098 763.660011,920.153374 C762.433258,918.94116 760.434387,918.94116 759.205094,920.153374 L752.672569,926.608477 C751.760759,927.509481 751.257867,928.709147 751.257867,929.981595 C751.257867,931.254044 751.760759,932.453709 752.672569,933.354713 C753.58438,934.255717 754.798434,934.752649 756.086144,934.752649 C757.376393,934.752649 758.587907,934.255717 759.499718,933.354713 L766.923734,926.021194 C767.1701,925.777747 767.568858,925.777747 767.815225,926.021194 C768.061592,926.26464 768.061592,926.658673 767.815225,926.902119 L760.391209,934.238148 C759.243192,935.37507 757.714195,936 756.086144,936 Z"
          />
        </g>
      </svg>
      Button text
    </div>
  </label>,
  <div
    className="styled__HelpTextCont-sc-3qf63z-4 iFNHeD"
  >
    Helper text
  </div>,
  <div
    className="styled__ImagesCont-sc-3qf63z-5 bOSahM"
  />,
]
`;
